### PR TITLE
fix(vanilla): fix unexpected cache in jotai-scope

### PR DESCRIPTION
**Pull request by yf-yang**
## Related Issues or Discussions

Fixes https://github.com/jotaijs/jotai-scope/issues/21

## Summary

Here `isSelfAtom` should not be used as it could lead to unexpected cache.

## Check List

- [x] `yarn run prettier` for formatting code and docs

### CODE REVIEW

**Comment by yf-yang** [comment]
The failed case can work with this patch.

**Comment by dai-shi** [comment]
Hm, okay, this is tricky, but alright.

Can you check this https://github.com/pmndrs/jotai/pull/2363#discussion_r1472011121 too?

**Comment by yf-yang** [comment]
Actually, I think `isSelfAtom` only matters when writing an atom (because we should write to the correct atom). For read(sub) purposes, since they are always connected via atom dependency, `isSelfAtom` does not matter from the correctness perspective (but maybe it could have some performance benefits)?

**Comment by dai-shi** [comment]
> but maybe it could have some performance benefits

`isSelfAtom` is not only for `jotai-scope`, so some kind of "correctness" is important.

File: src/vanilla/store.ts [hunk 1]
```diff
@@ -377,7 +377,7 @@ export const createStore = () => {
       // If all dependencies haven't changed, we can use the cache.
       if (
         Array.from(atomState.d).every(([a, s]) => {
-          if (isSelfAtom(atom, a)) {
+          if (atom === a) {
```
Comment (dai-shi) [inline src/vanilla/store.ts:380]:
```suggestion
          // we shouldn't use isSelfAtom. https://github.com/pmndrs/jotai/pull/2371
          if (a === atom) {
```

**Review by dai-shi — APPROVED** [review]
Thanks for your contribution! Looking forward to more suggestions.

**Pull request merged**